lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.h

```diff
@@ -140,24 +137,46 @@ class ProxyRenderDelegate : public MHWRender::MPxSubSceneOverride
     void _FilterSelection();
     void _UpdateSelectionStates();
 
-    const MayaUsdProxyShapeBase*  _proxyShape{ nullptr }; //!< DG proxy shape node
-    MDagPath                      _proxyDagPath;          //!< DAG path of the proxy shape (assuming no DAG instancing)
+    /*! \brief  Hold all data related to the proxy shape.
+
+        In addition to holding data read from the proxy shape, ProxyShapeData tracks when data read from the
+        proxy shape changes. For simple numeric types cache the last value read from _proxyShape & compare to
+        the current value. For complicated types we keep a version number of the last value we read to make 
+        fast comparisons.
+    */
+    class ProxyShapeData
+    {
+        const MayaUsdProxyShapeBase* const  _proxyShape{ nullptr };     //!< DG proxy shape node
+        const MDagPath                      _proxyDagPath;              //!< DAG path of the proxy shape (assuming no DAG instancing)
+        UsdStageRefPtr                      _usdStage;                  //!< USD stage pointer
+        size_t                              _excludePrimsVersion{ 0 };  //!< Last version of exluded prims used during render index populate
+        size_t                              _usdStageVersion{ 0 };      //!< Last version of stage used during render index populate
+    public:
+        ProxyShapeData(const MayaUsdProxyShapeBase* proxyShape, const MDagPath& proxyDagPath);
+        const MayaUsdProxyShapeBase* ProxyShape() const;
+        const MDagPath& ProxyDagPath() const;
+        UsdStageRefPtr UsdStage() const;
+        void UpdateUsdStage();
+        bool IsUsdStageUpToDate() const;
+        void UsdStageUpdated();
+        bool IsExcludePrimsUpToDate() const;
+        void ExcludePrimsUpdated();
+    };
+    std::unique_ptr<ProxyShapeData>          _proxyShapeData;
```

Comment:
You seem to assume everywhere that proxyShapeData exists...so why to make it dynamically allocated?

Reply:
I want to have a constructor to set the proxyShape pointer and MDagPath pointer. Those things need to get pulled out of the MObject passed into the ProxyRenderDelegate constructor. Basically if I made it a member I'd need to have the body of the ProxyRenderDelegate constructor appear in it's member initializers or give up the const-ness of some members of ProxyShapeData.